File: tests/function/inline/function_inline.ml
open Optitrust
open Target

(* let _ = Flags.check_validity := true *)

let _ = Run.script_cpp (fun _ ->

  let tf = cTopFunDef "test_fun" in
  !! Function_basic.inline ~body_mark:"bodyf" [tf;cCall "f"];
  !! Function_basic.inline [tf;cCall "g"];
  !! Function_basic.inline ~body_mark:"bodyh" [tf;cCall "h"];
  !! Function_basic.inline [tf;cCall "m"];
  !! Function_basic.inline ~body_mark:"bodyk" ~subst_mark:"substk" [tf;cCall "k"];

  (* FIXME:
  let tc = cTopFunDef "test_class_method" in

   (* Note: before inline any class method, all the members of the class should be made public! *)
  !! Record_basic.make_all_memebers_public [cTypDef "Test_method_inline"];

  !! Function_basic.inline ~body_mark:"bodyf1" [tc;cCall "f1"];
  !! Function_basic.inline ~body_mark:"bodyf1" [tc;cCall "f"];

  !! Function_basic.inline [tc;cCall "g"];
  !! Function_basic.inline ~body_mark:"bodyh" [tc;cCall "h"];
  !! Function_basic.inline [tc;cCall "m"];
  !! Function_basic.inline ~body_mark:"bodyk" [tc;cCall "k"];
  *)

  !! Trace.failure_expected (fun _e -> true) (fun () ->
    Function_basic.inline [cTopFunDef "test_nameclash"; cCall "f"]);

  !! Function.inline ~recurse:true [multi cCall ["recurse3"; "recurse2"]];
)

File: tests/function/inline/function_inline.cpp
int f(int x) {
  int a = x + x;
  return a + a;
}

int g(int x) {
  if (x > 0)
    return 1;
  else
    return 2;
}

int h(int x) {
  if (x > 0)
    return 1;
  return 2;
}

void m(int* p) {
  (*p)++;
}

int k(int a, int b) {
  return a + b;
}

void test_fun(){
  int x = 3;
  int y = f(x);
  int z = g(x);
  int u = h(x);
  int *q;
  m(q);
  int result;
  result = 10;
  // FIXME: bugged
  result = k(result, 4);
}


class Test_method_inline {
  private:
    int u;

  public:
    int f(int x){
      int a = x + x;
      return a + a;
    }

    int f1(){
      int a = u + u;
      return a + a;
    }

    int g(int x) {
      if (x > 0)
        return 1;
      else
        return 2;
    }

    int h(int x) {
      if (x > 0)
        return 1;
      return 2;
    }
    void m(int *x){
      (*x)++;
    }
    int k(int a, int b){
      return a + b;
    }
};

void test_class_method (){
  Test_method_inline c;
  int x = 3;

  int y = c.f(x);

  int y1 = c.f1();
  int z = c.g(x);
  int u = c.h(x);
  int *q;
  c.m(q);
  int result;
  result = 10;
  result = c.k(result, 4);
}

void test_nameclash() {
  int x = f(1);
  int y = f(2);
}

int recurse3() {
  return 0;
}

int recurse2() {
  int a = recurse3();
  int b = recurse3();
  // TODO: inline a/b, doesn't work yet
  return a + b;
}

int recurse1() {
  int a = recurse2();
  int b = recurse3();
  return a + b;
}
